### ISSUE TITLE
Template of `ExitHook` failed to resolve `{{workflow.failures}}`

### ISSUE THREAD

**Issue by jswxstw** [argoproj/argo-workflows#13245]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

Template of `ExitHook` failed to resolve `{{workflow.failures}}` as below: 
<img width="687" alt="image" src="https://github.com/argoproj/argo-workflows/assets/18543532/5124736d-5f6e-477c-a321-cf3e6c49b101">

The same template works well if using `onExit`, but it is `DEPRECATED`.
```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: workflow-exit-handler-demo
spec:
  entrypoint: main
  onExit: exit-handler
  templates:
  - name: main
    steps:
    - - name: hello
        template: hello
  - name: hello
    container:
      image: alpine:3.18
      command: [sh, -c]
      args: ["echo hello"]
  - name: exit-handler
    container:
      image: alpine:3.18
      command: [sh, -c]
      args: ["echo {{workflow.failures}}"]
```

### Version

latest

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: workflow-exit-handler-demo
spec:
  entrypoint: main
  hooks:
    exit:
      template: exit-handler
  templates:
  - name: main
    steps:
    - - name: fail
        template: fail
  - name: fail
    container:
      image: alpine:3.18
      command: [sh, -c]
      args: ["echo hello > /dev/termination-log;exit 1"]
  - name: exit-handler
    container:
      image: alpine:3.18
      command: [sh, -c]
      args: ["echo {{workflow.failures}}"]
```

### Logs from the workflow controller

```text
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```

### Logs from in your workflow's wait container

```text
kubectl logs -n argo -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

**Comment by tooptoop4** [comment]
maybe related https://github.com/argoproj/argo-workflows/issues/12330

**Comment by agilgur5** [comment]
@jswxstw can you clarify in the issue and PR (#13246) why you closed them?

It is very useful for historical purposes and for future readers who stumble upon them, and, as such, is a best practice we should always follow

**Comment by jswxstw** [comment]
> @jswxstw can you clarify in the issue and PR (#13246) why you closed them?
> 
> It is very useful for historical purposes and for future readers who stumble upon them, and, as such, is a best practice we should always follow

Closed as duplicate with #12330
